functions/lib/cla.js

```diff
@@ -1,74 +1,138 @@
 module.exports = Cla
 
+/**
+ * Returns CLA-related functions.
+ * @param db {FirebaseFirestore.Firestore}
+ */
 function Cla (db) {
-  async function isClaSigned (user, ref) {
-    if (!user || !user.email) {
-      console.log('email is not provided')
+  /**
+   * Returns a string that uniquely identifies the given identity object.
+   * @param identity {{type: string, value: string}}
+   * @returns {string|boolean}
+   */
+  function identityKey (identity) {
+    if (!identity) {
       return false
     }
+    if (!Object.prototype.hasOwnProperty.call(identity, 'type') ||
+      !Object.prototype.hasOwnProperty.call(identity, 'value')) {
+      console.warn('Identity must have at least a type, and a value')
+      return false
+    }
+    const lcValue = identity.value.toLowerCase()
+    return `${identity.type}:${lcValue}`
+  }
 
-    const email = user.email.toLowerCase()
-
-    try {
-      const claRef = db.collection('clas')
-      const clasByEmail = claRef.where('whitelist', 'array-contains', email)
-      let snapshot = await clasByEmail.get()
-
-      // Check if there is any CLA that signed for this email
-      let signed = !snapshot.empty
-
-      // TODO just for debugging
-      //  snapshot.forEach(doc => {
-      //      console.log(doc.id, ' => ', doc.toJson().whitelist);
-      //  })
-
-      // If there is no CLA signed by email, then check by domain
-      if (!signed) {
-        const domain = email.substring(email.lastIndexOf('@') + 1)
-        const clasByDomain = claRef.where('domain', '==', domain)
-        snapshot = await clasByDomain.get()
+  /**
+   * Given a snapshot of a newly created addendum, updates the whitelist in the
+   * parent agreement by replying all addendums in chronological order.
+   * @param {DocumentSnapshot} snapshot of a new addendum
+   * @returns {Promise}
+   */
+  async function updateWhitelist (snapshot) {
+    const newAddendumDoc = snapshot.data()
+    const agreementId = newAddendumDoc.agreementId
+    return db.collection('agreements').doc(agreementId).get()
+      .then(agreement => {
+        if (!agreement.exists) {
+          return Promise.reject(new Error('Agreement does not exist'))
+        }
+        return db.collection('addendums')
+          .where('agreementId', '==', agreementId)
+          .orderBy('dateSigned')
+          .get().then(function (query) {
```

Comment:
Instead of nesting promises you can use the format:
```
return db.collection('agreements').doc(agreementId).get()
      .then(agreement => {
            // do stuff
            return db.collection('addendums')
                  .where('agreementId', '==', agreementId)
                  .orderBy('dateSigned')
                  .get()
      })
      .then(query => {
            // do stuff
      })
      .catch(err => {
            // handle error
      })
```

Reply:
done. much better.